modules/nlu/src/backend/pipelines/intents/svm_classifier.ts

```diff
@@ -1,408 +1,94 @@
-    }
-
-    const bestOf3STD = math.std(predictions.slice(0, 3).map(p => p.confidence))
-    return bestOf3STD <= 0.03
+  } catch (e) {
+    throw new VError(e, `Error predicting intent for "${tokens.join(' ')}"`)
   }
+}
 
-  public get isLoaded(): boolean {
-    return !!Object.keys(this.l1PredictorsByContextName).length && !!this.l0Predictor
-  }
-
-  public async predict(tokens: string[], includedContexts: string[]): Promise<sdk.NLU.Intent[]> {
-    if (!this.isLoaded) {
-      throw new Error('No model loaded. Make sure you `load` your models before you call `predict`.')
-    }
-
-    if (!tokens.length) {
-      return []
-    }
-
-    if (!includedContexts.length) {
-      includedContexts = ['global']
+export const train = async function(
+  intentsWTokens: IntentDefinitionWithTokens[],
+  modelHash: string,
+  lang,
+  toolkit: typeof sdk.MLToolkit,
+  langProvider: LanguageProvider,
+  tfIdf: { [context: string]: _.Dictionary<_.Dictionary<number>> },
+  token2vec: Token2Vec,
+  realtime: typeof sdk.realtime,
+  realtimePayload: typeof sdk.RealTimePayload
+): Promise<Model[]> {
+  const notify = notifyProgress(realtime, realtimePayload)(identityProgress)
+  notify(0.1)
+
+  const contexts = getContextsFromIntentDefs(intentsWTokens)
+
+  const allPoints = await Promise.map(contexts, async context => {
+    const intentsForContext = intentsWTokens.filter(intent => intent.contexts.includes(context))
+
+    return {
+      context,
+      points: await getPointsForContext(intentsForContext, context, lang, langProvider, tfIdf, token2vec)
     }
+  })
 
-    const input = tokens.join(' ')
-
-    const l0Vec = await getSentenceFeatures({
-      lang: this.language,
-      doc: tokens,
-      docTfidf: this.l0Tfidf,
-      langProvider: this.languageProvider,
-      token2vec: this.token2vec
-    })
-
-    const l0Features = [...l0Vec, tokens.length]
-    const l0 = await this.predictL0Contextually(l0Features, includedContexts)
-
-    try {
-      debugPredict('prediction request %o', { includedContexts, input })
+  notify(0.2)
 
-      const predictions = _.flatten(
-        await Promise.map(includedContexts, async ctx => {
-          const l1Vec = await getSentenceFeatures({
-            lang: this.language,
-            doc: tokens,
-            docTfidf: this.l1Tfidf[ctx],
-            langProvider: this.languageProvider,
-            token2vec: this.token2vec
-          })
-          const l1Features = [...l1Vec, tokens.length]
-          const preds = await this.l1PredictorsByContextName[ctx].predict(l1Features)
-          const l0Conf = _.get(l0.find(x => x.label === ctx), 'confidence', 0)
+  // + 1 for global
+  const ctxLength = allPoints.map(x => x.context).length + 1
+  const ratioedProgress = getPayloadForInnerSVMProgress(ctxLength)
 
-          if (preds.length <= 0) {
-            return []
-          }
+  const l1Models = await createl1ModelsFromAllPoints(allPoints, modelHash, toolkit, notify, ratioedProgress(0))
```

Comment:
why not pass just an augmented instance of `notify` instead of passing a confusing progressFn (just reverse the arguments order)

Reply:
Yeah I know that's kind of weird, the thing is that the progress needs to progress during l1 training & l0 training. l1Models share the same indices so it's easy to use their index but l0 training progress needs to be offset by `l1Models.length`, once the offset is set, they can use their local index and then provide the progress value for the SVM.